File: microservice-architecture/resilience-patterns/python/test/test_resilience.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.resilience import (  # noqa: E402
    CircuitBreaker,
    CircuitOpenError,
    Result,
    ScriptedDependency,
    TimeoutError,
    call_with_timeout,
    retry,
)


class ResilienceTest(unittest.TestCase):
    def test_retry_eventually_succeeds(self) -> None:
        dependency = ScriptedDependency(
            [
                Result(error=RuntimeError("temporary failure")),
                Result(error=RuntimeError("temporary failure")),
                Result(value="OK"),
            ]
        )

        value, attempts = retry(3, dependency.call)
        self.assertEqual("OK", value)
        self.assertEqual(3, attempts)

    def test_timeout_fails_slow_dependency(self) -> None:
        dependency = ScriptedDependency([Result(value="SLOW_OK", delay_seconds=0.1)])
        with self.assertRaises(TimeoutError):
            call_with_timeout(0.01, dependency.call)

    def test_circuit_breaker_opens_after_repeated_failures(self) -> None:
        breaker = CircuitBreaker(2)
        dependency = ScriptedDependency(
            [
                Result(error=RuntimeError("dependency down")),
                Result(error=RuntimeError("dependency still down")),
                Result(value="RECOVERED"),
            ]
        )

        self.assertEqual("FALLBACK", breaker.execute(dependency.call, "FALLBACK"))
        self.assertEqual("FALLBACK", breaker.execute(dependency.call, "FALLBACK"))
        with self.assertRaises(CircuitOpenError):
            breaker.execute(dependency.call, "FALLBACK")

        breaker.reset()
        self.assertEqual("RECOVERED", breaker.execute(dependency.call, "FALLBACK"))


if __name__ == "__main__":
    unittest.main()

File: microservice-architecture/resilience-patterns/js/test/test_resilience.js
import {
  callWithTimeout,
  CircuitBreaker,
  CircuitOpenError,
  OperationTimeoutError,
  retry,
  ScriptedDependency
} from '../src/resilience.js'

function assertEqual(expected, actual, message) {
  if (expected !== actual) {
    throw new Error(`${message} expected=${expected} actual=${actual}`)
  }
}

const retryDependency = new ScriptedDependency([
  { error: new Error('temporary failure') },
  { error: new Error('temporary failure') },
  { value: 'OK' }
])
const retryResult = await retry(3, () => retryDependency.call())
assertEqual('OK', retryResult.value, 'retry result')
assertEqual(3, retryResult.attempts, 'retry attempts')

const timeoutDependency = new ScriptedDependency([{ value: 'SLOW_OK', delayMs: 100 }])
try {
  await callWithTimeout(10, () => timeoutDependency.call())
  throw new Error('timeout should have failed')
} catch (error) {
  if (!(error instanceof OperationTimeoutError)) {
    throw error
  }
}

const breaker = new CircuitBreaker(2)
const breakerDependency = new ScriptedDependency([
  { error: new Error('dependency down') },
  { error: new Error('dependency still down') },
  { value: 'RECOVERED' }
])
assertEqual('FALLBACK', await breaker.execute(() => breakerDependency.call(), 'FALLBACK'), 'first fallback')
assertEqual('FALLBACK', await breaker.execute(() => breakerDependency.call(), 'FALLBACK'), 'second fallback')

try {
  await breaker.execute(() => breakerDependency.call(), 'FALLBACK')
  throw new Error('circuit should be open')
} catch (error) {
  if (!(error instanceof CircuitOpenError)) {
    throw error
  }
}

breaker.reset()
assertEqual('RECOVERED', await breaker.execute(() => breakerDependency.call(), 'FALLBACK'), 'recovery')

console.log('resilience-patterns(js) tests passed')
